org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/IssueLocationsView.java

```diff
@@ -0,0 +1,402 @@
+        }
+        return new Object[] {new RootNode(sonarlintMarker, positions(p -> p.getMarkerId() == sonarlintMarker.getId()))};
+      } else {
+        return new Object[] {"No aditional locations associated with this issue"};
+      }
+    }
+
+    @Override
+    public Object[] getChildren(Object parentElement) {
+      if (parentElement instanceof RootNode) {
+        List<List<ExtraPosition>> flows = ((RootNode) parentElement).getFlows();
+        if (flows.size() > 1) {
+          AtomicInteger counter = new AtomicInteger(0);
+          return flows.stream().map(f -> f.size() > 1 ? new FlowRootNode("Flow " + counter.incrementAndGet(), f) : f.get(0)).toArray();
+        } else if (flows.size() == 1) {
+          return flows.get(0).toArray();
+        } else {
+          return new Object[0];
+        }
+      } else if (parentElement instanceof FlowRootNode) {
+        return ((FlowRootNode) parentElement).getPositions().toArray();
+      } else {
+        return new Object[0];
+      }
+    }
+
+    private List<ExtraPosition> positions(Predicate<? super ExtraPosition> filter) {
+      try {
+        return Arrays.asList(currentDoc.getPositions(MarkerUtils.SONARLINT_EXTRA_POSITIONS_CATEGORY))
+          .stream()
+          .map(p -> (ExtraPosition) p)
+          .filter(filter)
+          .collect(Collectors.toList());
+      } catch (BadPositionCategoryException e) {
+        SonarLintLogger.get().error("Unable to read positions", e);
+        return Collections.emptyList();
+      }
+    }
+
+    @Override
+    public Object getParent(Object element) {
+      return null;
+    }
+
+    @Override
+    public boolean hasChildren(Object element) {
+      return getChildren(element).length > 0;
+    }
+
+    @Override
+    public void dispose() {
+      // Do nothing
+    }
+
+    @Override
+    public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
+      // Do nothing
+    }
+
+  }
+
+  private static class LocationsViewer extends TreeViewer {
+
+    public LocationsViewer(Tree tree) {
+      super(tree);
+      setAutoExpandLevel(ALL_LEVELS);
+      setUseHashlookup(true);
+    }
+  }
+
+  private static class LocationsTreeLabelProvider extends LabelProvider {
+
+    @Override
+    public String getText(Object element) {
+      if (element instanceof RootNode) {
+        return ((RootNode) element).getRoot().getAttribute(IMarker.MESSAGE, "No message");
+      } else if (element instanceof FlowRootNode) {
+        return ((FlowRootNode) element).getLabel();
+      } else if (element instanceof String) {
+        return (String) element;
+      }
+      return ((ExtraPosition) element).getMessage();
+    }
+
+    @Override
+    public Image getImage(Object element) {
+      if (element instanceof RootNode) {
+        return SonarLintImages.IMG_ISSUE;
+      }
+      return super.getImage(element);
+    }
+
+  }
+
+  public void setInput(@Nullable IMarker sonarlintMarker) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)